subsys/net/ip/net_tls.c

```diff
@@ -0,0 +1,735 @@
+				  struct net_tls_credential *ca_cert)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	int err = mbedtls_x509_crt_parse(&context->tls->ca_chain,
+					 ca_cert->buf, ca_cert->len);
+	if (err != 0) {
+		return -EINVAL;
+	}
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+
+	return 0;
+}
+
+static void tls_set_ca_chain(struct net_context *context)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	mbedtls_ssl_conf_ca_chain(&context->tls->config,
+				  &context->tls->ca_chain, NULL);
+
+	mbedtls_ssl_conf_authmode(&context->tls->config,
+				  MBEDTLS_SSL_VERIFY_REQUIRED);
+
+	mbedtls_ssl_conf_cert_profile(&context->tls->config,
+				      &mbedtls_x509_crt_profile_default);
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+}
+
+static int tls_set_psk(struct net_context *context,
+		       struct net_tls_credential *psk,
+		       struct net_tls_credential *psk_id)
+{
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
+	mbedtls_ssl_conf_psk(&context->tls->config,
+			     psk->buf, psk->len,
+			     (const unsigned char *)psk_id->buf,
+			     psk_id->len - 1);
+#endif
+
+	return 0;
+}
+
+static int tls_mbedtls_set_credentials(struct net_context *context)
+{
+	struct net_tls_credential *credential;
+	sec_tag_t tag;
+	int i, err;
+	bool ca_cert_present = false;
+
+	for (i = 0; i < context->tls->options.sec_tag_list.sec_tag_count; i++) {
+		tag = context->tls->options.sec_tag_list.sec_tags[i];
+		credential = NULL;
+
+		while ((credential = credential_next_get(tag, credential))
+				!= NULL) {
+			switch (credential->type) {
+			case NET_TLS_CREDENTIAL_CA_CERTIFICATE:
+			{
+				err = tls_add_ca_certificate(context,
+							     credential);
+				if (err < 0) {
+					return err;
+				}
+
+				ca_cert_present = true;
+
+				break;
+			}
+
+			case NET_TLS_CREDENTIAL_PSK:
+			{
+				struct net_tls_credential *psk_id =
+					credential_get(
+						tag, NET_TLS_CREDENTIAL_PSK_ID);
+
+				if (!psk_id) {
+					return -ENOENT;
+				}
+
+				err = tls_set_psk(context, credential, psk_id);
+				if (err < 0) {
+					return err;
+				}
+
+				break;
+			}
+
+			case NET_TLS_CREDENTIAL_PSK_ID:
+				/* Ignore PSK ID - it will be used together with PSK */
+				break;
+
+			default:
+				return -EINVAL;
+			}
+		}
+	}
+
+	if (ca_cert_present) {
+		tls_set_ca_chain(context);
+	}
+
+	return 0;
+}
+
+void net_tls_init(void)
+{
+	static const unsigned char drbg_seed[] = "zephyr";
+	int i, ret;
+
+	for (i = 0; i < ARRAY_SIZE(credentials); i++) {
+		credentials[i].type = NET_TLS_CREDENTIAL_UNUSED;
+	}
+
+#if defined(CONFIG_NET_PRECONFIGURE_TLS_CREDENTIALS)
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	net_tls_credential_add(NET_TLS_DEFAULT_CA_CERTIFICATE_TAG,
+			       NET_TLS_CREDENTIAL_CA_CERTIFICATE,
+			       ca_certificate, sizeof(ca_certificate));
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
+	net_tls_credential_add(NET_TLS_DEFAULT_PSK_TAG,
+			       NET_TLS_CREDENTIAL_PSK,
+			       client_psk, sizeof(client_psk));
```

Comment:
Ah, so here how it works. If user defined CONFIG_NET_PRECONFIGURE_TLS_CREDENTIALS, then that file now has to contain definitions of both ca_certificate and client_psk/client_psk_id. That's not good.